File: app/admin/transaction/direct_purchase.rb
ActiveAdmin.register DirectPurchase do

  menu parent: "Transaction"

  permit_params :purchase_date, :supplier, :product, :quantity, :unit_cost, :commission_rate, 
  :total_unit_cost, :liens_cost, :insurance_cost, :storage_cost, :penalty_cost, :witholding_tax, 
  :commission_amount, :payable_gross, :payable_net, :reference_number

  config.sort_order = "updated_at_desc"

  index do
    selectable_column
    column :purchase_date
    column "Reference #" do |direct_purchase|
      direct_purchase.reference_number
    end
    column :supplier
    column :product
    column :quantity
    column :unit_cost do |direct_purchase|
      number_to_currency(direct_purchase.unit_cost, unit: "₱")
    end
    column "Total Commision" do |direct_purchase|
      number_to_currency(direct_purchase.commission_rate, unit: "₱")
    end
    column :total_unit_cost do |direct_purchase|
      number_to_currency(direct_purchase.total_unit_cost, unit: "₱")
    end
    column :witholding_tax do |direct_purchase|
      number_to_currency(direct_purchase.witholding_tax, unit: "₱")
    end
    column :commission_amount do |direct_purchase|
      number_to_currency(direct_purchase.commission_amount, unit: "₱")
    end
    column "Liens" do |direct_purchase|
      number_to_currency(direct_purchase.liens_cost, unit: "₱")
    end
    column "Insurance" do |direct_purchase|
      number_to_currency(direct_purchase.insurance_cost, unit: "₱")
    end
    column "Insurance" do |direct_purchase|
      number_to_currency(direct_purchase.storage_cost, unit: "₱")
    end
    column "Penalty" do |direct_purchase|
      number_to_currency(direct_purchase.penalty_cost, unit: "₱")
    end
    column :payable_gross do |direct_purchase|
      number_to_currency(direct_purchase.payable_gross, unit: "₱")
    end
    column :payable_net do |direct_purchase|
      number_to_currency(direct_purchase.payable_net, unit: "₱")
    end
    column :updated_at
    actions dropdown: true
  end

  show title: :reference_number do
    panel "Direct Purchase Details" do
      attributes_table_for direct_purchase do
        row :purchase_date
        row "Reference #" do
          direct_purchase.reference_number
        end
        row :supplier
        row :product
        row :quantity
      end
      panel "Costs" do
        attributes_table_for direct_purchase do
          row :unit_cost  do |direct_purchase|
            number_to_currency(direct_purchase.unit_cost, unit: "₱")
          end
          row :commission_rate do |direct_purchase|
            number_to_currency(direct_purchase.commission_rate, unit: "₱")
          end
          row :total_unit_cost do |direct_purchase|
            number_to_currency(direct_purchase.total_unit_cost, unit: "₱")
          end
          row :witholding_tax do |direct_purchase|
            number_to_currency(direct_purchase.witholding_tax, unit: "₱")
          end
          row "Total commission" do |direct_purchase|
            number_to_currency(direct_purchase.commission_amount, unit: "₱")
          end
        end
        panel "Misc Fees" do
          attributes_table_for direct_purchase do
            row "Liens" do |direct_purchase|
              number_to_currency(direct_purchase.liens_cost, unit: "₱")
            end
            row "Insurance" do |direct_purchase|
              number_to_currency(direct_purchase.insurance_cost, unit: "₱")
            end
            row "Storage" do |direct_purchase|
              number_to_currency(direct_purchase.storage_cost, unit: "₱")
            end
            row "Penalty" do |direct_purchase|
              number_to_currency(direct_purchase.penalty_cost, unit: "₱")
            end
          end
        end
      end
      panel "Payable" do
        attributes_table_for direct_purchase do
          row "Gross" do |direct_purchase|
            number_to_currency(direct_purchase.payable_gross, unit: "₱")
          end
          row "Net" do |direct_purchase|
            number_to_currency(direct_purchase.payable_net, unit: "₱")
          end
        end
      end
      attributes_table_for direct_purchase do
        row :created_at
        row :updated_at
      end
    end
    active_admin_comments
  end
  
  form do |f|
    panel "Direct Purchase Formulas" do
      ul do
        li "Total Unit Cost = Unit Cost + Commission Rate"
        li "Commssion Amount = Quantity * Commission Rate"
        li "Payable Gross = Quantity + Total Unit Cost"
        li "Payable Net = Gross Payable - Misc Fees - Tax"
      end
    end
    panel "Direct Purchase" do
      f.inputs "Details" do
        f.input :purchase_date, as: :datepicker
        f.input :reference_number, label: "Reference #"
        f.input :supplier, :as => :select, :collection => Supplier.all.map {|supplier| [supplier.supplier_name, supplier.supplier_name]}, :input_html_options => { :class => 'chosen'}
        f.input :product, :as => :select, :collection => Product.all.map {|product| [product.item_name, product.item_name, product.item_name]}, :input_html_options => { :class => 'chosen'}
        f.input :quantity, :input_html => { :value => 0 }
      end
      panel "Costs" do
        f.inputs do
          f.input :unit_cost, :input_html => { :value => 0 }
          f.input :commission_rate, :input_html => { :value => 0 }
          f.input :total_unit_cost, :input_html => { :readonly => true, :value => 0 }
          f.input :witholding_tax, :input_html => { :value => 0 }
          f.input :commission_amount, label: "Total commission", :input_html => { :readonly => true, :value => 0 }
        end
        f.inputs "Misc Fees" do
          f.input :liens_cost, label: "Liens", :input_html => { :value => 0 }
          f.input :insurance_cost, label: "Insurance", :input_html => { :value => 0 }
          f.input :storage_cost, label: "Storage", :input_html => { :value => 0 }
          f.input :penalty_cost, label: "Penalty", :input_html => { :value => 0 }
        end
      end
      f.inputs "Payable" do
        f.input :payable_gross, label: "Gross", :input_html => { :readonly => true, :value => 0 }
        f.input :payable_net, label: "Net", :input_html => { :readonly => true, :value => 0 }
      end
    end
    f.actions
  end

end

File: app/assets/javascripts/direct_purchase.js
$(document).ready(function(){

	//auto-fill unit cost of product in direct purchase
  $("#direct_purchase_product").change(function() {
  	$("#direct_purchase_unit_cost").val($(this).val());
  	alert($(this).val());
	});

  //compute for total unit cost when unit cost is changed
	$("#direct_purchase_unit_cost").change(function() {
		var unit_cost = parseFloat($("#direct_purchase_unit_cost").val()),
				commission_rate = parseFloat($("#direct_purchase_commission_rate").val()),
				total_unit_cost = unit_cost + commission_rate;

  	$("#direct_purchase_total_unit_cost").val(total_unit_cost)
	});

	//compute for total unit cost and commission amount when commission rate is changed
	$("#direct_purchase_commission_rate").change(function() {
		var unit_cost = parseFloat($("#direct_purchase_unit_cost").val()),
				commission_rate = parseFloat($("#direct_purchase_commission_rate").val()),
				quantity = parseFloat($("#direct_purchase_quantity").val()),
				total_unit_cost = unit_cost + commission_rate;
				commission_amount = quantity * commission_rate;

  	$("#direct_purchase_total_unit_cost").val(total_unit_cost);
  	$("#direct_purchase_commission_amount").val(commission_amount);
	});

	//compute for commission amount and payable gross when quantity is changed
	$("#direct_purchase_quantity").change(function() {
		var quantity = parseFloat($("#direct_purchase_quantity").val()),
				commission_rate = parseFloat($("#direct_purchase_commission_rate").val()),
				total_unit_cost = parseFloat($("#direct_purchase_total_unit_cost").val()),
				commission_amount = quantity * commission_rate,
				payable_gross = quantity * total_unit_cost;

  	$("#direct_purchase_commission_amount").val(commission_amount);
  	$("#direct_purchase_payable_gross").val(payable_gross);
	});

	//compute for total unit cost when total unit cost is hovered
	$("#direct_purchase_total_unit_cost").mouseenter(function() {
		var unit_cost = parseFloat($("#direct_purchase_unit_cost").val()),
				commission_rate = parseFloat($("#direct_purchase_commission_rate").val()),
				total_unit_cost = unit_cost + commission_rate;

  	$("#direct_purchase_total_unit_cost").val(total_unit_cost);
	});

	//compute for total unit cost when total unit cost is focused
	$("#direct_purchase_total_unit_cost").focus(function() {
		var unit_cost = parseFloat($("#direct_purchase_unit_cost").val()),
				commission_rate = parseFloat($("#direct_purchase_commission_rate").val()),
				total_unit_cost = unit_cost + commission_rate;

  	$("#direct_purchase_total_unit_cost").val(total_unit_cost);
	});

	//compute for commission amount when commission amount is hovered
	$("#direct_purchase_commission_amount").mouseenter(function() {
		var quantity = parseFloat($("#direct_purchase_quantity").val()),
				commission_rate = parseFloat($("#direct_purchase_commission_rate").val()),
				commission_amount = quantity * commission_rate;

  	$("#direct_purchase_commission_amount").val(commission_amount);
	});

	//compute for commission amount when commission amount is focused
	$("#direct_purchase_commission_amount").focus(function() {
		var quantity = parseFloat($("#direct_purchase_quantity").val()),
				commission_rate = parseFloat($("#direct_purchase_commission_rate").val()),
				commission_amount = quantity * commission_rate;

  	$("#direct_purchase_commission_amount").val(commission_amount);
	});

  //compute for payable gross when payable gross is hovered
	$("#direct_purchase_payable_gross").mouseenter(function() {
		var quantity = parseFloat($("#direct_purchase_quantity").val()),
				total_unit_cost = parseFloat($("#direct_purchase_total_unit_cost").val()),
				payable_gross = quantity * total_unit_cost;

  	$("#direct_purchase_payable_gross").val(payable_gross);
	});

	//compute for payable gross when payable gross is focused
	$("#direct_purchase_payable_gross").focus(function() {
		var quantity = parseFloat($("#direct_purchase_quantity").val()),
				total_unit_cost = parseFloat($("#direct_purchase_total_unit_cost").val()),
				payable_gross = quantity * total_unit_cost;

  	$("#direct_purchase_payable_gross").val(payable_gross);
	});

	//compute for payable net when payable net is hovered
	$("#direct_purchase_payable_net").mouseenter(function() {
		var witholding_tax = parseFloat($("#direct_purchase_witholding_tax").val()),
				misc_fees = parseFloat($("#direct_purchase_liens_cost").val()) + parseFloat($("#direct_purchase_storage_cost").val()) + 
				parseFloat($("#direct_purchase_insurance_cost").val()) + parseFloat($("#direct_purchase_penalty_cost").val()),
				payable_gross = parseFloat($("#direct_purchase_payable_gross").val())
				payable_net = payable_gross - (witholding_tax + misc_fees);
				
  	$("#direct_purchase_payable_net").val(payable_net);
  });

  //compute for payable net when payable net is focused
	$("#direct_purchase_payable_net").focus(function() {
		var witholding_tax = parseFloat($("#direct_purchase_witholding_tax").val()),
				misc_fees = parseFloat($("#direct_purchase_liens_cost").val()) + parseFloat($("#direct_purchase_storage_cost").val()) + 
				parseFloat($("#direct_purchase_insurance_cost").val()) + parseFloat($("#direct_purchase_penalty_cost").val()),
				payable_gross = parseFloat($("#direct_purchase_payable_gross").val())
				payable_net = payable_gross - (witholding_tax + misc_fees);
				
  	$("#direct_purchase_payable_net").val(payable_net);
  });

	// compute for payable net when witholding tax is changed
	$("#direct_purchase_witholding_tax").change(function() {
		var witholding_tax = parseFloat($("#direct_purchase_witholding_tax").val()),
				misc_fees = parseFloat($("#direct_purchase_liens_cost").val()) + parseFloat($("#direct_purchase_storage_cost").val()) + 
				parseFloat($("#direct_purchase_insurance_cost").val()) + parseFloat($("#direct_purchase_penalty_cost").val()),
				payable_gross = parseFloat($("#direct_purchase_payable_gross").val())
				payable_net = payable_gross - (witholding_tax + misc_fees);
				
  	$("#direct_purchase_payable_net").val(payable_net);
  });

  // compute for payable net when liens cost is changed
	$("#direct_purchase_liens_cost").change(function() {
		var witholding_tax = parseFloat($("#direct_purchase_witholding_tax").val()),
				misc_fees = parseFloat($("#direct_purchase_liens_cost").val()) + parseFloat($("#direct_purchase_storage_cost").val()) + 
				parseFloat($("#direct_purchase_insurance_cost").val()) + parseFloat($("#direct_purchase_penalty_cost").val()),
				payable_gross = parseFloat($("#direct_purchase_payable_gross").val())
				payable_net = payable_gross - (witholding_tax + misc_fees);
				
  	$("#direct_purchase_payable_net").val(payable_net);
  });

  // compute for payable net when storage cost is changed
	$("#direct_purchase_storage_cost").change(function() {
		var witholding_tax = parseFloat($("#direct_purchase_witholding_tax").val()),
				misc_fees = parseFloat($("#direct_purchase_liens_cost").val()) + parseFloat($("#direct_purchase_storage_cost").val()) + 
				parseFloat($("#direct_purchase_insurance_cost").val()) + parseFloat($("#direct_purchase_penalty_cost").val()),
				payable_gross = parseFloat($("#direct_purchase_payable_gross").val())
				payable_net = payable_gross - (witholding_tax + misc_fees);
				
  	$("#direct_purchase_payable_net").val(payable_net);
  });

  // compute for payable net when insurance cost is changed
	$("#direct_purchase_insurance_cost").change(function() {
		var witholding_tax = parseFloat($("#direct_purchase_witholding_tax").val()),
				misc_fees = parseFloat($("#direct_purchase_liens_cost").val()) + parseFloat($("#direct_purchase_storage_cost").val()) + 
				parseFloat($("#direct_purchase_insurance_cost").val()) + parseFloat($("#direct_purchase_penalty_cost").val()),
				payable_gross = parseFloat($("#direct_purchase_payable_gross").val())
				payable_net = payable_gross - (witholding_tax + misc_fees);
				
  	$("#direct_purchase_payable_net").val(payable_net);
  });

  // compute for payable net when penalty cost is changed
	$("#direct_purchase_penalty_cost").change(function() {
		var witholding_tax = parseFloat($("#direct_purchase_witholding_tax").val()),
				misc_fees = parseFloat($("#direct_purchase_liens_cost").val()) + parseFloat($("#direct_purchase_storage_cost").val()) + 
				parseFloat($("#direct_purchase_insurance_cost").val()) + parseFloat($("#direct_purchase_penalty_cost").val()),
				payable_gross = parseFloat($("#direct_purchase_payable_gross").val())
				payable_net = payable_gross - (witholding_tax + misc_fees);
				
  	$("#direct_purchase_payable_net").val(payable_net);
  });

}); 